src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,692 @@
+  unordered_map<Node, unsigned, NodeHashFunction> visited;
+  unordered_map<Node, unsigned, NodeHashFunction>::iterator it;
+
+  visit.push(expr);
+  while (!visit.empty())
+  {
+    Node n = visit.top();
+    visit.pop();
+    it = visited.find(n);
+    if (it == visited.end())
+    {
+      visited[n] = 0;
+      visit.push(n);
+      for (Node child : n) visit.push(child);
+    }
+    else if (it->second == 0)
+    {
+      Kind k = n.getKind();
+      switch (k)
+      {
+        case kind::CONST_BITVECTOR:
+        {
+          unsigned bwval;
+          Integer val = n.getConst<BitVector>().getValue();
+          for (bwval = 0; val > 0; val = val.divByPow2(1), ++bwval)
+            ;
+          visited[n] = bwval;
+          break;
+        }
+
+        case kind::BITVECTOR_EXTRACT:
+        {
+          unsigned w = utils::getSize(n);
+          visited[n] = w >= visited[n[0]] ? visited[n[0]] : w;
+          Assert(visited[n] <= visited[n[0]]);
+          break;
+        }
+
+        case kind::BITVECTOR_ZERO_EXTEND:
+        {
+          visited[n] = visited[n[0]];
+          break;
+        }
+
+        case kind::BITVECTOR_MULT:
+        {
+          unsigned w = 0;
+          for (unsigned i = 0, nc = n.getNumChildren(); i < nc; ++i)
+            w += visited[n[i]];
+          if (w > utils::getSize(n)) return 0; /* overflow */
+          visited[n] = w;
+          break;
+        }
+
+        case kind::BITVECTOR_CONCAT:
+        {
+          unsigned w;
+          Node zero = utils::mkZero(utils::getSize(n[0]));
+          if (n[0] == zero)
+          {
+            w = visited[n[1]]
+                /* add bw of children n[2] to n[n.getNumChildren()-1] */
+                + utils::getSize(n) - utils::getSize(n[0])
+                - utils::getSize(n[1]);
+          }
+          else
+          {
+            w = visited[n[0]]
+              /* add bw of children n[1] to n[n.getNumChildren()-1] */
+              +  utils::getSize(n) - utils::getSize(n[0]);
+          }
+          visited[n] = w;
+          break;
+        }
+
+        case kind::BITVECTOR_UDIV_TOTAL:
+        case kind::BITVECTOR_OR:
+        case kind::BITVECTOR_NOR:
+        case kind::BITVECTOR_XOR:
+        case kind::BITVECTOR_XNOR:
+        case kind::BITVECTOR_AND:
+        case kind::BITVECTOR_NAND:
+        case kind::BITVECTOR_SHL:
+        case kind::BITVECTOR_LSHR:
+        case kind::BITVECTOR_ASHR:
+        {
+          unsigned wmax = 0;
+          for (unsigned i = 0, nc = n.getNumChildren(); i < nc; ++i)
+            if (visited[n[i]] > wmax) wmax = visited[n[i]];
+          visited[n] = wmax;
+          break;
+        }
+
+        case kind::BITVECTOR_UREM_TOTAL:
+        {
+          visited[n] = visited[n[1]];
+          break;
+        }
+
+        case kind::BITVECTOR_PLUS:
+        case kind::BITVECTOR_SUB:
+        {
+          unsigned w = 0, wmax = visited[n[0]];
+          for (unsigned i = 1, nc = n.getNumChildren(); i < nc; ++i)
```

Comment:
No, that's not true. `wmax` would be initialized with 20, and since none of the other children is greater or equal than 20, the determined bit-width would be 21, which is correct. Starting from `i = 0` would have the same effect:

```
unsigned w = 0, wmax = 0;
for (unsigned i = 0, nc = n.getNumChildren(); i < nc; ++i)
{
  if (visited[n[i]] > wmax) wmax = visited[n[i]];
}
```

Maybe this is easier to read, so I might change this.

Reply:
You are right, I must've missed the initialization of wmax. The proposed change would be easier to read though. Btw. I think that the calculated number of bits is too big: consider an addition of four 20 bit numbers. The current code would say that this needs 23 bits but it actually only needs 22 bits: if the addition is a+b+c+d, then a+b and c+d are each at most 21 bits and thus the whole expressions is at most 22 bits. I think the actual bitwidth is something like wmax + log2(w) (+/- 1).

Reply:
This is now computed via determining the required bit-width of the sum of all max values (of the minimum bit-width of the operands). E.g., if you have one operand with min bit-width 2 and one with min bit-width 4 the result would be 5 (`11 + 1111 = 10010` -> min bit-width: 5). This is probably as close to the minimum required bit-width as it gets. I applied the same tactic to MULT.